Resumption wire protocol

espadolini opened the pull request:
This PR adds the implementation for the inner wire protocol that's going to be used by SSH connection resumption (as defined in [RFD 0150](https://github.com/gravitational/teleport/blob/master/rfd/0150-ssh-connection-resumption.md)).

It's not yet used in actual code, but it's tested for compliance by both checking that it works correctly as a pipe with `nettest.TestConn` (without exercising resumption), and by checking that breaking the underlying connection and then attaching a new one lets the connection continue unimpeded.

Implements [RFD 0150](https://github.com/gravitational/teleport/blob/master/rfd/0150-ssh-connection-resumption.md).
Part of #30415.

espadolini committed:
WIP: resumption wire protocol and in-memory pipe

espadolini committed:
more test cases for pipe

espadolini committed:
testing and error returning

espadolini committed:
remove in-memory pipe implementation for now

espadolini committed:
don't check the remote address (it should be done earlier)

espadolini committed:
Avoid allocating on every iteration of the write loop

espadolini committed:
add ResumableConn constructor

espadolini committed:
refactoring

espadolini commented on lib/resumption/resumable.go:
```diff
@@ -0,0 +1,327 @@
+	eg, ctx := errgroup.WithContext(context.Background())
+	context.AfterFunc(ctx, requestStop)
+
+	eg.Go(func() error {
```
I split the two internal goroutines into external functions, which made things a bit clearer too.

espadolini committed:
address pr comments

espadolini commented:
@GavinFrazar @r0mant ping

GavinFrazar commented on lib/resumption/resumable.go:
```diff
@@ -0,0 +1,334 @@
+type Conn struct {
+	managedConn
+
+	// attached is non-nil iff there's an underlying connection attached;
+	// calling it should eventually result in the connection becoming detached,
+	// signaled by the field becoming nil.
+	attached func()
```
nit: might a better name be `detachUnderlying`?

espadolini replied:
Renamed to `requestDetach`.

GavinFrazar commented on lib/resumption/resumable.go:
```diff
@@ -0,0 +1,334 @@
+		r.cond.Broadcast()
+		r.mu.Unlock()
+
+		_, _ = nc.Write([]byte(errorTagUvarint))
```
should we use a write deadline here?

espadolini replied:
That's covered by the `handshakeWatchdog` timer that closes the connection unless the handshake completes within 5 seconds (`handshakeTimeout`).

GavinFrazar commented on lib/resumption/resumable.go:
```diff
@@ -0,0 +1,334 @@
+			return trace.Wrap(err, "writing frame data")
+		}
+
+		r.mu.Lock()
+		sentPosition += frameAck
+		peerPosition += len64(frameData)
+		r.mu.Unlock()
```
why does this need to lock?

espadolini replied:
The read loop used to interact with one of these in a previous implementation but it's not needed anymore, thank you for noticing.

espadolini committed:
attached -> requestDetach

espadolini committed:
remove useless locked section

espadolini committed:
Close the conn if either loops exit

espadolini committed:
only close after the read loop is completed

This also makes the test wait on close, to confirm that the underlying
connection is also closed (shortly) after the real connection is closed.

GavinFrazar commented on lib/resumption/resumable.go:
```diff
@@ -0,0 +1,339 @@
+	defer nc.Close()
+
+	if !firstConn {
+		for !r.remoteClosed && r.requestDetach != nil {
+			r.requestDetach()
+			r.cond.Wait()
+		}
```
is this dead code? Below we defer a function that sets `r.requestDetach = nil`

espadolini replied:
Before returning we will set `r.requestDetach` to `nil`, but there might be more than one call of `runResumeV1Unlocking` running at the same time; this loop is specifically to ask the currently attached run to detach and leave the connection to us.

Remember that at this point we're holding the lock on `r`, but `r.cond.Wait()` will unlock it and let some other goroutine proceed on it until a `r.cond.Broadcast()`, and we, in fact, broadcast after setting `requestDetach` to `nil`.

GavinFrazar commented on lib/resumption/resumable.go:
```diff
@@ -0,0 +1,339 @@
+
+	handshakeWatchdog := time.AfterFunc(handshakeTimeout, func() { nc.Close() })
+	defer handshakeWatchdog.Stop()
+
+	if !firstConn {
+		errC := make(chan error, 1)
+		go func() {
+			_, err := nc.Write(binary.AppendUvarint(nil, sentPosition))
+			errC <- err
+		}()
+		var err error
+		peerPosition, err = binary.ReadUvarint(ncReader)
+		if err != nil {
+			return trace.Wrap(err, "reading peer receive position")
+		}
+		err = <-errC
+		if err != nil {
+			return trace.Wrap(err, "writing receive position")
+		}
+	}
+
+	r.mu.Lock()
+	if minPos, maxPos := r.sendBuffer.start, r.sendBuffer.end; peerPosition < minPos || maxPos < peerPosition {
+		// incompatible resume position, mark as remotely closed since we can't
+		// ever continue from this; this also includes receiving an errorTag
+		// (since that's too big of a position to reach legitimately)
+		r.remoteClosed = true
+		r.cond.Broadcast()
+		r.mu.Unlock()
+
+		_, _ = nc.Write([]byte(errorTagUvarint))
+		return trace.BadParameter("got incompatible resume position (%v, expected %v to %v)", peerPosition, minPos, maxPos)
+	}
+	if r.sendBuffer.start != peerPosition {
+		r.sendBuffer.advance(peerPosition - r.sendBuffer.start)
+		r.cond.Broadcast()
+	}
+	r.mu.Unlock()
+
+	handshakeWatchdog.Stop()
```
if it's not too onerous, can you refactor this into a `handshake` function, just to make it easier to understand what it's doing?

espadolini commented on lib/resumption/resumable.go:
```diff
@@ -0,0 +1,339 @@
+			}
+
+			if r.remoteClosed {
+				r.mu.Unlock()
+				return trace.ConnectionProblem(errBrokenPipe, "connection closed by peer")
+			}
+
+			if r.localClosed {
+				r.mu.Unlock()
+
+				_, _ = nc.Write([]byte(errorTagUvarint))
+				return trace.ConnectionProblem(net.ErrClosed, "connection closed")
```
Won't `errors.Is` also unwrap the error if necessary? I don't think anything expects either, so we can wrap a `nil` error with `ConnectionProblem` or just `trace.Wrap` the semantic errors with additional messages.

espadolini replied:
Replaced every synthesized error with a `trace.Wrap(net.ErrClosed, ...)`, since that will get ignored when logging (by going through `utils.IsOKNetworkError`) and the only important errors that we might care about (as they'd reveal logical problems) are still `trace.BadParameter` anyway.

espadolini commented on lib/resumption/resumable.go:
```diff
@@ -0,0 +1,339 @@
+		}
+	} else if r.requestDetach != nil || r.remoteClosed || r.localClosed || r.receiveBuffer.end > 0 || r.sendBuffer.start > 0 {
+		r.mu.Unlock()
+		panic("firstConn for resume V1 is not actually unused")
```
The first connection is special at the resumption protocol level, it's the connection in which the connection was first established, and in that connection we skip sending 0 and receiving 0 from the other end. It's possible to resume a connection that wasn't closed, that was already detached, and through which there was no data sent, but that still goes through the "resume existing connection" path and at the wire level both sides need to exchange the stream position.

Even more crucially, it's possible for both sides to disagree on whether or not the connection has been "used"; if only side A has sent some data but it hasn't made through it to side B, side B will have a completely clean state but side A will have some data ready to send, thus failing that check.

espadolini committed:
minor rework
- handshake in a separate function
- clearer logic for exiting on localClosed

espadolini committed:
explanation for the asymmetry between read and write

espadolini commented on lib/resumption/resumable.go:
```diff
@@ -0,0 +1,339 @@
+	context.AfterFunc(ctx, requestStop)
+
+	eg.Go(func() error {
+		defer nc.Close()
```
I added some explanations about why we close the underlying connection when exiting from the read loop (and, crucially, why we don't when exiting from the write loop).

GavinFrazar replied:
thanks!

espadolini committed:
trace.ConnectionProblem considered harmful

espadolini commented on lib/resumption/resumable_test.go:
I'll add some tests for an incompatible resume handshake and for oversized frames.

espadolini committed:
test naming

fspmarshall commented on lib/resumption/resumable.go:
```diff
@@ -0,0 +1,348 @@
+	return peerPosition, nil
+}
+
+func runResumeV1Read(r *Conn, nc byteReaderReader, stopRequested *atomic.Bool) error {
```
This and `runResumeV1Write` would both benefit a _lot_ from inline comments.  These functions are really really dense and a lot of what they are doing is a bit opaque.  Detailed comments would go a long way toward making these functions more reviewable/maintainable.

espadolini replied:
added

espadolini committed:
document and simplify consts

espadolini committed:
copyright year

espadolini commented on lib/resumption/resumable.go:
```diff
@@ -0,0 +1,334 @@
+
+const handshakeTimeout = 5 * time.Second
+
+const (
+	errorTag        = ^uint64(0)
+	errorTagUvarint = "\xff\xff\xff\xff\xff\xff\xff\xff\xff\x01"
+)
+
+const maxFrameSize = 128 * 1024
```
added

espadolini commented on lib/resumption/resumable.go:
```diff
@@ -0,0 +1,334 @@
+		for !r.remoteClosed && r.attached != nil {
+			r.attached()
+			r.cond.Wait()
+		}
```
I don't think it makes too much sense to do it like that because wakeups can happen for all sorts of reasons - a goroutine reading one byte at a time `n` times will generate `n` wakeups and trigger the warning even though there's nothing wrong with the detaching procedure - and counting how many different instances of `requestDetach` the detach loop has encountered is really awkward because `func()` is not comparable (except to `nil`).

Do you have ideas on how to do that?

espadolini replied:
Added a time-based check to the detach loop; we're going to notice if a deadlock completely breaks SSH, but some change that makes detaching really slow will only result in hiccups, so logging that seems appropriate. We do the same thing for slow TLS handshakes and slow sqlite transactions.

espadolini committed:
Warn on slow detach

espadolini committed:
document runResumeV1Write and runResumeV1Read

espadolini committed:
make stopRequested non-atomic, protected by mu
